File: old/parse-scenes.py
import json
import os
from typing import NamedTuple, List, Dict, Any, Tuple

class Vector3(NamedTuple[float, float, float]):
    x: float
    y: float
    z: float


class Spawn(NamedTuple[int, int, int, int, Vector3, Vector3]):
    spawnId: int
    configId: int
    level: int
    poseId: int
    pos: Vector3
    rot: Vector3


class Scene(NamedTuple[int, int, int, Vector3, List[Spawn]]):
    sceneId: int
    groupId: int
    blockId: int
    pos: Vector3
    spawns: List[Spawn]


def parse_vector3(d: Dict[str, Any]) -> Tuple[float, float, float]:
    return d['x'], d['y'], d['z']



def parse_spawn(d: Dict[str, Any]) -> Tuple[int, int, int, int, Tuple[float, float, float], Tuple[float, float, float]]:
    return d['monsterId'], d['configId'], d['level'], d['poseId'], parse_vector3(d['pos']), parse_vector3(d['rot'])


def parse_scene(d: Dict[str, Any]) -> Scene:
    return d['sceneId'], d['groupId'], d['blockId'], parse_vector3(d['pos']), [parse_spawn(s) for s in d['spawns']]


with open('data/Spawns.json') as data:
    d = json.load(data)
scenes: List[Scene] = [Scene(*parse_scene(s)) for s in d]

File: src/parse-scenes.ts
import * as fs from 'fs'

interface Vector3 {
    x: number;
    y: number;
    z: number;
}

interface Spawn {
    monsterId: number;
    configId: number;
    level: number;
    poseId: number;
    pos: Vector3;
    rot: Vector3;
}

interface Scene {
    sceneId: number;
    groupId: number;
    blockId: number;
    pos: Vector3;
    spawns: Spawn[];
}

function Vector3(d: any): Vector3 {
    return {
        x: d.x,
        y: d.y,
        z: d.z
    };
}

function parseSpawn(d: any): Spawn {
    return {
        monsterId: d.monsterId,
        configId: d.configId,
        level: d.level,
        poseId: d.poseId,
        pos: Vector3(d.pos),
        rot: Vector3(d.rot)
    };
}

function parseScene(d: any): Scene {
    return {
        sceneId: d.sceneId,
        groupId: d.groupId,
        blockId: d.blockId,
        pos: Vector3(d.pos),
        spawns: d.spawns.map(parseSpawn)
    };
}

const data = fs.readFileSync('./data/Spawns.json');
const d = JSON.parse(data.toString());
let scenes: Scene[] = d.map(parseScene);

interface ParsedScene {
    [SceneId: number]: number[];
}

interface ParsedSpawn {
    [MonsterId: number]: number[][]
}

let parsedScenes: ParsedScene = {

};

let parsedSpawns: ParsedSpawn = {};

scenes.forEach(scene => {
    if (!parsedScenes[scene.sceneId])
        parsedScenes[scene.sceneId] = [];
    parsedScenes[scene.sceneId].push(scene.sceneId);
    scene.spawns.forEach(spawn => {
        if (!parsedSpawns[spawn.monsterId])
            parsedSpawns[spawn.monsterId] = [];
        parsedSpawns[spawn.monsterId].push([spawn.pos.x, spawn.pos.y, spawn.pos.z]);
    });
});

fs.writeFileSync('./parsed/parsedScenes.json', JSON.stringify(parsedScenes));
fs.writeFileSync('./parsed/parsedSpawns.json', JSON.stringify(parsedSpawns));